File: packages/agent-sdk-python/lazynext_agent/memory.py
"""
Memory management for the Lazynext AI Agent Agent Python SDK.

Provides an in-process key-value store with summarisation capabilities,
mirroring the TypeScript MemoryManager API.
"""

from __future__ import annotations

import json
from datetime import datetime, timezone
from typing import Generic, TypeVar

T = TypeVar("T")


class MemoryManager:
    """In-process key-value store with summarisation.

    Useful for:
    - Caching agent preferences across queries
    - Storing intermediate results the agent may re-use
    - Building a working set that can be summarised for the LLM
    """

    def __init__(self) -> None:
        self._store: dict[str, tuple[object, float]] = {}

    def remember(self, key: str, value: object) -> None:
        """Store a value under the given key.

        Overwrites any existing entry with the same key.
        """
        self._store[key] = (value, datetime.now(timezone.utc).timestamp())

    def recall(self, key: str, default: T | None = None) -> T | None:
        """Retrieve a previously stored value.

        Returns *default* when the key does not exist.
        """
        entry = self._store.get(key)
        if entry is None:
            return default
        return entry[0]  # type: ignore[return-value]

    def forget(self, key: str) -> bool:
        """Remove a single key from the store.

        Returns ``True`` if the key existed, ``False`` otherwise.
        """
        if key in self._store:
            del self._store[key]
            return True
        return False

    def has(self, key: str) -> bool:
        """Check whether a key exists in the store."""
        return key in self._store

    def keys(self) -> list[str]:
        """Return all keys currently in the store."""
        return list(self._store.keys())

    def summarize(self) -> str:
        """Generate a human-readable summary of all stored entries."""
        if not self._store:
            return "(empty)"

        lines: list[str] = []
        for key, (value, _ts) in self._store.items():
            val = value if isinstance(value, str) else json.dumps(value)
            lines.append(f"- {key}: {val}")
        return "\n".join(lines)

    def to_dict(self) -> dict[str, dict[str, object]]:
        """Return all entries as a plain dict (suitable for JSON serialisation)."""
        result: dict[str, dict[str, object]] = {}
        for key, (value, ts) in self._store.items():
            result[key] = {
                "value": value,
                "timestamp": datetime.fromtimestamp(ts, tz=timezone.utc).isoformat(),
            }
        return result

    def clear(self) -> None:
        """Remove all entries from the store."""
        self._store.clear()

    @property
    def size(self) -> int:
        """Return the number of entries currently stored."""
        return len(self._store)

File: packages/agent-sdk/src/memory.ts
/**
 * Memory management for the Lazynext AI Agent Agent SDK.
 *
 * Provides a key-value store with summarisation capabilities that
 * persists for the lifetime of the {@link Lazynext AI AgentAgent} instance.
 * The remote gateway also maintains its own memory store; call
 * {@link Lazynext AI AgentAgent.getMemory} to retrieve the server-side state.
 *
 * ```ts
 * import { MemoryManager } from "@lazynext/agent-sdk";
 *
 * const mem = new MemoryManager();
 * mem.remember("target_lufs", -16);
 * const lufs = mem.recall<number>("target_lufs");
 * const summary = mem.summarize();
 * ```
 *
 * @module agent-sdk/memory
 */

/** A single entry in the in-memory store. */
interface MemoryEntry {
	/** Stored value. */
	value: unknown;
	/** Milliseconds timestamp of when stored. */
	timestamp: number;
}

/**
 * In-process key-value store with summarisation.
 *
 * The memory manager is useful for:
 * - Caching agent preferences across queries
 * - Storing intermediate results the agent may re-use
 * - Building a working set that can be summarised for the LLM
 */
export class MemoryManager {
	private store: Map<string, MemoryEntry>;

	constructor() {
		this.store = new Map();
	}

	/**
	 * Store a value under the given key.
	 *
	 * Overwrites any existing entry with the same key.
	 * The value is stored by reference — use `structuredClone` if
	 * you need a deep copy.
	 */
	remember(key: string, value: unknown): void {
		this.store.set(key, {
			value,
			timestamp: Date.now(),
		});
	}

	/**
	 * Retrieve a previously stored value.
	 *
	 * Returns `undefined` when the key does not exist.
	 * The caller should cast the result to the expected type.
	 *
	 * ```ts
	 * const speed = mem.recall<number>("clip_speed");
	 * ```
	 */
	recall<T = unknown>(key: string): T | undefined {
		const entry = this.store.get(key);
		return entry ? (entry.value as T) : undefined;
	}

	/**
	 * Remove a single key from the store.
	 *
	 * Returns `true` if the key existed, `false` otherwise.
	 */
	forget(key: string): boolean {
		return this.store.delete(key);
	}

	/**
	 * Check whether a key exists in the store.
	 */
	has(key: string): boolean {
		return this.store.has(key);
	}

	/**
	 * Return all keys currently in the store.
	 */
	keys(): string[] {
		return Array.from(this.store.keys());
	}

	/**
	 * Generate a human-readable summary of all stored entries.
	 *
	 * Each key-value pair is formatted as a bullet point.  Values are
	 * serialised with `JSON.stringify`.
	 */
	summarize(): string {
		if (this.store.size === 0) return "(empty)";

		const lines: string[] = [];
		for (const [key, entry] of this.store) {
			const val =
				typeof entry.value === "string"
					? entry.value
					: JSON.stringify(entry.value);
			lines.push(`- ${key}: ${val}`);
		}
		return lines.join("\n");
	}

	/**
	 * Return all entries as a plain object suitable for JSON serialisation.
	 *
	 * Timestamps are ISO-8601 strings.
	 */
	toJSON(): Record<string, { value: unknown; timestamp: string }> {
		const result: Record<
			string,
			{ value: unknown; timestamp: string }
		> = {};
		for (const [key, entry] of this.store) {
			result[key] = {
				value: entry.value,
				timestamp: new Date(entry.timestamp).toISOString(),
			};
		}
		return result;
	}

	/**
	 * Remove all entries from the store.
	 */
	clear(): void {
		this.store.clear();
	}

	/**
	 * Return the number of entries currently stored.
	 */
	get size(): number {
		return this.store.size;
	}
}
